File: app/src/main/java/com/anthony/deepl/openl/model/TranslationRequest.java
package com.anthony.deepl.openl.model;

import com.google.gson.annotations.SerializedName;

import com.anthony.deepl.openl.manager.LanguageManager;

import java.text.BreakIterator;
import java.util.ArrayList;
import java.util.List;

public class TranslationRequest {

    @SerializedName("jsonrpc")
    private String mJsonRpc = "2.0";

    @SerializedName("method")
    private String mMethod = "LMT_handle_jobs";

    @SerializedName("params")
    private TranslationRequestParams mParams;

    private List<Integer> mLineBreakPositions;

    public TranslationRequest(String sentence, String fromLanguage, String toLanguage, List<String> userPreferredLanguages) {
        List<TranslationRequestJob> jobList = new ArrayList<>();
        BreakIterator sentenceIterator = BreakIterator.getSentenceInstance(LanguageManager.getLocaleFromLanguageValue(fromLanguage, null));
        int sentenceStart, sentenceEnd;
        int jobCount = 0;
        mLineBreakPositions = new ArrayList<>();

        sentenceIterator.setText(sentence);
        for (sentenceStart = sentenceIterator.first(), sentenceEnd = sentenceIterator.next(); sentenceEnd != BreakIterator.DONE; sentenceStart = sentenceEnd, sentenceEnd = sentenceIterator.next()) {
            TranslationRequestJob job = new TranslationRequestJob();
            String detectedSentence = sentence.substring(sentenceStart, sentenceEnd);
            job.setSentence(detectedSentence);
            jobList.add(job);
            if (detectedSentence.contains("\n")) {
                mLineBreakPositions.add(jobCount);
            }
            jobCount++;
        }

        mParams = new TranslationRequestParams();
        mParams.setJobList(jobList);
        mParams.setRequestLanguages(new TranslationRequestLanguage(fromLanguage, toLanguage, userPreferredLanguages));
    }

    public List<Integer> getLineBreakPositions() {
        return mLineBreakPositions;
    }
}

class TranslationRequestParams {

    @SerializedName("priority")
    private short mPriority = -1;

    @SerializedName("jobs")
    private List<TranslationRequestJob> mJobs;

    @SerializedName("lang")
    private TranslationRequestLanguage mLanguages;

    void setJobList(List<TranslationRequestJob> jobs) {
        mJobs = jobs;
    }

    void setRequestLanguages(TranslationRequestLanguage languages) {
        mLanguages = languages;
    }

}

class TranslationRequestJob {

    @SerializedName("kind")
    private String mJobKind = "default";

    @SerializedName("raw_en_sentence")
    private String mSentence = "LMT_handle_jobs";

    void setSentence(String sentence) {
        mSentence = sentence;
    }

}

class TranslationRequestLanguage {

    @SerializedName("source_lang_user_selected")
    private String mSourceLanguage;

    @SerializedName("target_lang")
    private String mTargetLanguage;

    @SerializedName("user_preferred_langs")
    private List<String> mPreferredLanguages;

    public TranslationRequestLanguage(String sourceLanguage, String targetLanguage, List<String> preferredLanguages) {
        mSourceLanguage = sourceLanguage;
        mTargetLanguage = targetLanguage;
        mPreferredLanguages = preferredLanguages;
    }

}

File: app/src/main/java/com/anthony/deepl/openl/model/TranslationRequest.kt
package com.anthony.deepl.openl.model

import com.anthony.deepl.openl.manager.LanguageManager
import com.squareup.moshi.Json

import java.text.BreakIterator
import java.util.ArrayList

data class TranslationRequest(
        private val sentence: String,
        private val fromLanguage: String,
        private val toLanguage: String,
        private val userPreferredLanguages: List<String>,
        @field:Json(name="jsonrpc") private val jsonRpc: String = "2.0",
        @field:Json(name="method") private val method: String = "LMT_handle_jobs") {


    @field:Json(name = "params") private val mParams: TranslationRequestParams
    val lineBreakPositions: MutableList<Int>

    init {
        val jobList = ArrayList<TranslationRequestJob>()
        val sentenceIterator = BreakIterator.getSentenceInstance(LanguageManager.getLocaleFromLanguageValue(fromLanguage, null))
        var sentenceStart: Int
        var sentenceEnd: Int
        var jobCount = 0
        lineBreakPositions = ArrayList()

        sentenceIterator.setText(sentence)
        sentenceStart = sentenceIterator.first()
        sentenceEnd = sentenceIterator.next()
        while (sentenceEnd != BreakIterator.DONE) {
            val detectedSentence = sentence.substring(sentenceStart, sentenceEnd)
            jobList.add(TranslationRequestJob(sentence = detectedSentence))
            if (detectedSentence.contains("\n")) {
                lineBreakPositions.add(jobCount)
            }
            jobCount++
            sentenceStart = sentenceEnd
            sentenceEnd = sentenceIterator.next()
        }

        mParams = TranslationRequestParams(
                jobs = jobList,
                languages = TranslationRequestLanguage(fromLanguage, toLanguage, userPreferredLanguages))
    }
}

internal data class TranslationRequestParams(
        @field:Json(name="priority") val priority: Short = -1,
        @field:Json(name="jobs") val jobs: List<TranslationRequestJob>?,
        @field:Json(name="lang") val languages: TranslationRequestLanguage?
)

internal data class TranslationRequestJob(
        @field:Json(name="kind") val jobKind: String = "default",
        @field:Json(name="raw_en_sentence") val sentence: String = "LMT_handle_jobs"
)

internal data class TranslationRequestLanguage(
        @field:Json(name="source_lang_user_selected") val sourceLanguage: String,
        @field:Json(name="target_lang") val targetLanguage: String,
        @field:Json(name="user_preferred_langs") val preferredLanguages: List<String>
)